MSC2162: Signaling Errors at Bridges (#2162)
matrix-org/matrix-doc: proposals/2162-signaling-errors-at-bridges.md

# Signaling Errors at Bridges

Sometimes bridges just silently swallow messages and other events. This proposal
enables bridges to communicate that something went wrong and gives clients the
option to give feedback to their users.

> ## Proposal

Reviewer: The proposal does not cover how bridges de-flag errors (eventual success in sending a message). I am assuming they redact their original error event.

Reviewer 2: I believe this proposal is only handling the case of when the bridge gives up trying to send.

Reviewer: that sounds sub-par tbh. We'd need a retry mechanism so that users aren't left stranded, or at the very least support redaction as a way to indicate clearing of the error.

Author: If there is a good chance that a message will eventually be delivered, I don't think it belongs in this proposal. We should try to unify that case with a general “delivery delay notification” solution for the whole Matrix universe so the work has to be done only once. I am currently writing a bit about what I have in mind about those “delivery delay notifications” and there can discussion about that as well. (Also not quite sure where to have it then.) In the case of a message not being delivered with a high probability and just backing off in rare circumstances, redacting a permanent error might be adequate.

Until now I assumed the error is final and there is no retry, just a manual resend of the message. Could a bridge get the redaction and refetch the original event? Or might it be possible to simulate a resend with a no-op edit? If there is no satisfying way already, one could of course add another event type which is ignored by everyone but the bridge.

Reviewer: Homeservers are expected to keep retrying the appservice until it comes back alive, but that can easily be hours or even days before the service responds. Most bridges nowadays have a condition for received messages where ti just drops messages which are too old, but between the time the bridge went down and the time the message was ignored the user's message was not delivered without notification.

Limiting the scope of the proposal to just fatal errors doesn't really help with communicating the bridge's status because there's many more temporary failures that people expect to hear about due to the nature of realtime communications. It's bad enough we already get complaints when it takes more than 10 seconds to send a message through 4 different points of failure to the remote network.

Reviewer 2: I am coming round to the idea of sending a temporary failure *PDU* for things which have failed to send and are in a retry queue of some kind. Redacting that would imply it's been sent.

Separately there is a question of if this proposal should cover how the user can indicate they want to retry a message.

Reviewer: Hopefully coming around enough to give the OK 😇 

I'd be uncomfortable with this going into the spec if it only communicated permanent failures, because permanent failures are rare.

> ## Proposal

Reviewer: The proposal doesn't explain who sends the error event. Is it the `sender_localpart`? a user in the namespace? Both of these options have consequences which make them non-ideal: the `sender_localpart` might not be in the room in the case of a puppet bridge and a user in the namespace is an undefined thing: anyone can claim they are in a namespace.

Reviewer: I'm failing to see the correlation between this and authorization for bridges (I also don't know what proposal that is). Bridges have a namespace of users and a `sender_localpart`: which user is supposed to send it?

Author: One user representing the bridge does send the message. It depends on who is in the room, so the answer is both of them.

We can't simply say the bridge bot user as it is sometimes not joined e.g. in 1:1 conversations. Then the virtual user of your communication partner does represent the bridge and it should send the bridge error.

Reviewer: This causes problems (as does the regex later on) because clients won't be able to do sanity checking on errors. They don't have a concept of bridges or appservices, and would be unable to see that `@alice:example.org` doesn't have appropriate permissions over the Freenode IRC bridge.

> ## Proposal

Reviewer: The proposal should cover how much we care about random users impersonating bridges or bridges lying about their namespaces, and how we protect against that if we do care (we should).

Bridges might come into a situation where there is nothing more they can do to
successfully deliver an event to the foreign network they are connected to. Then

> they should be able to inform the originating group of the event about this

Reviewer: ```suggestion
they should be able to inform the originating room of the event about this
```

delivery error.

> This document proposes the addition of a new PDU event with type

Reviewer: ```suggestion
This document proposes the addition of a new room event with type
```

> `m.bridge_error`. It is sent by the bridge and references an event previously

Reviewer 3: Maybe worth mentioning that a bridge should never send a `m.bridge_error` message in response to a `m.bridge_error`?

Author: The response of an unaware bridge would be some moderately fun recursion with `m.unknown_event`. So I think yes, good point.

sent in a room, by that marking it as “failed to deliver” for all users of a
bridge. The new event type utilizes reference aggregations ([MSC

> 1849](https://github.com/matrix-org/matrix-doc/blob/matthew/msc1849/proposals/1849-aggregations.md))

Reviewer 2: Heads up that @ara4n is rewriting this proposal, but I don't think it will affect this one https://github.com/matrix-org/matrix-doc/pull/2154/files

to establish the relation to the event its delivery it is marking as failed.
There is no need for a new endpoint as the existing `/send` endpoint will be
utilized.

Additional information contained in the event are the name of the bridged
network (e.g. “Discord” or “Telegram”) and a regex¹ describing the affected

> users (e.g. `@discord_*:example.org`). This regex should be similar to the one

Reviewer 3: Is it common for a failure to only affect a subset of a bridge's users?  Can you give an example of a situation where this might occur?

Author: I don't think so. The intent is to identify the users of the bridge at all as each bridge may have an arbitrary username namespace.

Reviewer 3: OK, so the common usecase for this field would be to identify all of the bridge's users.

> users (e.g. `@discord_*:example.org`). This regex should be similar to the one

Reviewer 2: I want to jump in and say my preference is to keep it as a simple glob expression, rather than having to define the rules around regexes. I doubt we plan to do more than simple matching?

> users (e.g. `@discord_*:example.org`). This regex should be similar to the one

Reviewer: We need to define what kind of glob it is then, as we use two types in Matrix. Also if it's not a regex, don't call it a regex.

Being able to reuse the namespace from the registration is a strong argument for supporting proper regex here (which is better defined than globs).

> users (e.g. `@discord_*:example.org`). This regex should be similar to the one

Reviewer 3: > (which is better defined than globs).

FWIW, my impression is the opposite.  There are several different syntaxes for regexes.  But personally, I don't really care which one is picked, as long as it's well defined.

any Application Service uses for marking its reserved user namespace. By
providing this information clients can inform their users who in the room was
affected by the error and for which network the error occurred.

There are some common reasons why an error occurred. These are encoded in the
`reason` attribute and can contain the following types:

> * `m.event_not_handled` Generic error type for when an event can not be handled

Reviewer: let's not do that: the convention within the spec is to have `m.whatever` as your error, not `m.error.whatever`.

  by the bridge. It is used as a fallback when there is no other more specific
  reason.

> * `m.event_too_old` When the foreign network does not support timestamp

Reviewer: tbh as a bridge operator I'd rather drop the message on the floor than bother with signalling this error. I personally already run my bridges such that anything older than 15 minutes doesn't even get sent to the bridges to reduce traffic, useless processing, and context in the foreign network. Although there's the occasional question about why some messages don't get sent, I don't think it's worth spending the homeserver's time to advertise the failure.

Given these errors are associated with individual events, and time differences between events usually only happen after a major homeserver problem, this is just asking for a self-inflicted denial of service when the bridge is supposed to send thousands of errors into rooms. It's not very often that one or two events are too old - most of the time it's batches of hours worth of conversation.

tldr: I don't think we need this error code as it leads to self-inflicted denial of service.

>   massaging like Matrix does, a message will – with enough time passed – fall

Reviewer: ftr Matrix does not support massaging - it was in a draft specification, but was removed due to complexities around what the timestamp argument should do with respect to the DAG and more complicated endpoints.

  out of its original context. In this case the bridge might decide that the
  event is too old and emit this error.

* `m.foreign_network_error` The bridge was doing its job fine, but the foreign
  network permanently refused to handle the event.

> * `m.unknown_event` The bridge is not able to handle events of this type.

Reviewer: Of the type or format (or both)? A bridge might support text messages, but not HTML, or it might support stickers, but not emotes. The language here implies that a bridge has to support all of `m.room.message`.

Additionally, with the usual goal of a bridge to be instant messaging, I don't think it's fair to ever raise this error. The event types bridges wouldn't be able to send are the same event types it is unlikely to ever be able to send (server ACLs, room aliases, etc). On the other hand, if the bridge is able to raise this error for subtypes/formats of events it doesn't understand, this could be used to flag that a kick/ban was not carried over. 

Can we put some recommendations in here that bridges shouldn't send this error for events outside of its use case (ie: for instant messaging, don't signal errors on ACLs, aliases, and custom event types).

Author: Subtypes are fine, I think I clarified it now in the new version.

Let's assume a bridge that doesn't know about stickers yet. It would be very nice to have a sticker sent in that room trigger an `m.unknown_event` then.

I am more in favor of whitelisting some events by having the bridge recognize those event and not acting on them at all. Everything that is completely unknown would still trigger this. I don't expect there to be many false positives in practice, but if so, the UI for this type of error could be reduced to something less noisy.

> 

Reviewer 3: These seem to all be permanent errors.  At the risk of feature-creeping, is there any scope for temporary errors to indicate that messages may be delayed? e.g. the foreign network seems to be down at the moment, but the bridge will keep retrying for a while. (And if so, we'd need a mechanism for indicating when the error has been resolved, which could be done by redacting the `m.bridge_error` message)

Author: See above :)

(I had that in my [initial GSoC proposal](https://nasnotfound.de/nextcloud/index.php/s/YsgbRR7544N4fEp). Meanwhile it seems temporary failure is a bigger thing and quite different.)

Nothing prevents multiple bridge error events to relate to the same event. This
should be pretty common as a room can be bridged to more than one network at a
time.

The need for this proposal arises from a gap between the Matrix network and
other foreign networks it bridges to. Matrix with its eventual consistency is
unique in having a message delivery guarantee. Because of this property there is

> no need in the Matrix network itself to model the failure of message delivery.

Reviewer: Arguably there is for some obscure cases. For instance, if you set up a bridge in a read only room, the bridge might not be able to post messages. However, the foreign network likely doesn't support indicating failure so the room is the next best option to flag a potential problem. How it does that starts stepping into very questionable territory (extensible profiles, auth rules which support a limited set of per-user events, etc).

Reviewer 4: I'd probably de-scope this for the sake of this MSC landing any time soon.

Reviewer 4: But if we are de-scoping, the MSC text needs to be updated to reflect that and the edge cases that, like @turt2live mentioned, can still be encountered.

Author: I am still not quite sure if I properly get what you are saying, but I think it is now discussed in the [Rights management](https://github.com/V02460/matrix-doc/blob/msc2162/proposals/2162-signaling-errors-at-bridges.md#rights-management) section.

This need only arises for interactions with foreign networks where message
delivery might fail. This proposal extends Matrix to be aware of these error
cases.

Additionally there might be some operational restrictions of bridges which might
make it necessary for them to refrain from handling an event, e.g. when hitting
memory limits. In this case the new event type can be used as well.

### Example

```
{
    "type": "m.room.bridge_error",
    "content": {
        "network: "Discord",

>         "affected_users": "@discord_*:example.org",

Reviewer 2: Potentially this could be an array, contain a set of strings/glob expressions?

Author: In the current implementation I am simply or'ing all of the expressions the AS is returning into one. So not strictly necessary I think (as long as the chosen grammar is capable enough).

Reviewer: generally properties which sound like arrays should be arrays. +1 to an array here.

>         "reason": "m.event_too_old",

Reviewer 2: Having tried to implement this, I feel like a `reason_body` couldn't hurt for additional explanation. It's not very multi-language friendly, but there are error cases where we could do with further explanation.

For instance, I send a foreign network error because I couldn't connect an IRC client for a user. I need to also inform the reason for the failure is because of that.

Author: Would really prefer the alternative. I think it's very reasonable to come up with a set of common errors that cover most usecases. If we must we can add subtyping like `m.general_error.special.most_specialized`, which is a scheme that nicely combines backwards compatibility and extensibility.

Reviewer 2: Ooh, yeah that could work.

        "m.relates_to": {
            "rel_type": "m.reference",
            "event_id": "$some:event.id"
        }
    }
}
```

\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\
¹ Or similar – see *Security Considerations*

## Tradeoffs

Without this proposal, bridges could still inform users in a room that a
delivery failed by simply sending a plain message event from a bot account. This
possibility carries the disadvantage of conveying no special semantic meaning
with the consequence of clients not being able to adapt their presentation.

A fixed set of error types might be too restrictive to express every possible
condition. An alternative would be a free-form text for an error message. This
brings the problems of less semantic meaning and a requirement for
internationalization with it. In this proposal a generic error type is provided
for error cases not considered in this MSC.

## Potential issues

When the foreign network is not the cause of the error signaled but the bridge
itself (maybe under load), there might be an argument that responding to failed
messages increases the pressure.

> 

Reviewer 3: Another potential issue is that this doesn't convey any error information if messages failed to send due to the bridge being down completely (as the bridge is unable to send the error messages).

Author: When the bridge comes back online, it will receive the missed events from the HS, so they might be handled after all. This would be only temporary and by that explicitly not covered by this proposal. The big thing to tackle here would be a mechanism to [signal delivery delays](https://github.com/matrix-org/matrix-doc/issues/2094) which would add to the core Matrix network as well.

Reviewer: I still think this is something that should be addressed or at least mentioned in the proposal. If the homeserver cannot send the event to the bridge, it should send an error event on its behalf (which the bridge can later redact).

Reviewer 4: Seems like it is mentioned here? https://github.com/matrix-org/matrix-doc/pull/2162/files#diff-3fc0af60441d4268c3ff475f9e03fb4cR192

any Application Service uses for marking its reserved user namespace. By
providing this information clients can inform their users who in the room was
affected by the error and for which network the error occurred.

There are some common reasons why an error occurred. These are encoded in the
`reason` attribute and can contain the following types:

* `m.event_not_handled` Generic error type for when an event can not be handled
  by the bridge. It is used as a fallback when there is no other more specific
  reason.

* `m.event_too_old` A message will – with enough time passed – fall out of its
  original context. In this case the bridge might decide that the event is too
  old and emit this error.

* `m.foreign_network_error` The bridge was doing its job fine, but the foreign
  network permanently refused to handle the event.

* `m.unknown_event` The bridge is not able to handle events of this type. It is
  totally legitimate to “handle” an event by doing nothing and not throwing this
  error. It is at the discretion of the bridge author to find a good balance
  between informing the user and preventing unnecessary spam. Throwing this

>   error only for some subtypes of an event if fine.

Reviewer 4: ```suggestion
  error only for some subtypes of an event is fine.
```

* `m.bridge_unavailable` The homeserver couldn't reach the bridge.

* `m.no_permission` The bridge wanted to handle an event, but didn't have the
  permission to do so.

The bridge error can provide a `time_to_permanent` field. If this field is

> present it gives the time in seconds one has to wait before declaring the bridge

Reviewer 4: Most of the Matrix spec uses milliseconds instead of seconds.

error as permanent. As long as an error is younger than this time, the client
can expect the possibility of the error being revoked. If a bridge error is

> permanent, it should not be revoked anymore. In addition, the field may also

Reviewer 4: It might be easier to have 0 as "permanent instantly" and the lack of the field being "never"? That keeps the field to a single datatype.

Just thinking that in [Dendrite](https://github.com/matrix-org/dendrite) we use Golang which, like other typed languages, requires that we declare the type of a JSON field before we receive it. It's possible to specify multiple types, but it's a bit of a [faff](https://stackoverflow.com/questions/46279048/encoding-decoding-multi-type-fields-in-json-golang).

Plus, having the lack of a `time_to_permanent` field meaning "will never be permanent" sounds more intuitive to me than "permanent immediately".

This is an example of how the new bridge error might look:

```
{
    "type": "m.bridge_error",
    "content": {
        "network: "Discord",
        "affected_users": "@discord_*:example.org",
        "reason": "m.bridge_unavailable",
        "time_to_permanent": 900,
        "m.relates_to": {
            "rel_type": "m.reference",
            "event_id": "$some:event.id"
        }
    }
}
```

\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\
¹ Or similar – see *Security Considerations*

### Retries and error revocation

Providing a way to retry a failed message delivery gives the sender control over
the importance of her message. An extra procedure for a retry is necessary as
the message might have been delivered to some users (those not on the bridge)
and this would produce duplicate messages for them.

A retry request is posted by the client to the room for all bridges to see it,
referencing the original event. By inspecting the sender of all related
`m.bridge_error` events, under all bridges the correct one can find out that it
is responsible. The responsible bridge re-fetches the original event and retries
to deliver it.

> A successful retry should be communicated by revoking (not redacting) the

Reviewer 4: I assume the reason we redact is because we still want the original error information to be given to clients (i.e. which bridge was it, when did the error happen etc.

However that wasn't clear until I read a few paragraphs down. If this is the reason, could you briefly mention it here?

original error that made the retry necessary. Revocation is done by an event
with the type `m.bridge_error_revoke` which references the original event. The
error(s) having a sender of the same bridge as the revocation event are

> considered revoked. Clients can show a revoke error e.g. as “Delivered to

Reviewer 4: ```suggestion
considered revoked. Clients can show a revocation message e.g. as “Delivered to
```

Discord at 14:52.” besides the original event.

> On an unsuccessful retry the bridge may edit the errors content to reflect the

Reviewer 4: ```suggestion
On an unsuccessful retry the bridge may edit the error's content to reflect the
```

new state, e.g. because the type of error changed or to communicate the new
time.

Example of the new retry events:

```
{
    "type": "m.bridge_retry",
    "content": {
        "m.relates_to": {
            "rel_type": "m.reference",
            "event_id": "$original:event.id"
        }
    }
}
```

```
{
    "type": "m.bridge_error_revoke",
    "content": {
        "m.relates_to": {
            "rel_type": "m.reference",
            "event_id": "$original:event.id"
        }
    }
}
```

Overview of the relations between the different event types:

```

>  ________________     _____________________

Reviewer 4: ```suggestion
                           m.references
 ________________     _____________________
```

|                |   |                     |
| Original Event |-+-|    Bridge Error     |
|________________| | |_____________________|
                   |  _____________________
                   | |                     |
                   +-|    Retry Request    |

>       m.references | |_____________________|

Reviewer 4: ```suggestion
                   | |_____________________|
```

                   |  _____________________
                   | |                     |
                   +-| Bridge Error Revoke |
                     |_____________________|
```

A retry might not make much sense for every kind of error e.g. retrying
`m.unknown_event` will probably result in the same error again. Clients may
choose to disable retry options for those cases, but it is not restricted
otherwise.

### Special case: Unavailable bridge

In the case the bridge is down or otherwise disconnected from the homeserver, it
naturally has no way to inform its users about the unavailability. In this case
the homeserver can stand in as an agent for the bridge and answer requests in

> its responsibility.

Reviewer 4: ```suggestion
its absence.
```

changes to The Spec would be necessary before following the probably more
appropriate route here.

A new event type `m.bridge_error_revoke` is introduced for revoking a bridge
error. Alternatively it could be considered to redact the bridge error event,
which would eliminate the need for the revocation event and would make this
proposal a little simpler. The disadvantage of this approach is the missing
transparency and context of who had which information at which point in time.
This additional information should make for a better user experience.

## Potential issues

When the foreign network is not the cause of the error signaled but the bridge
itself (maybe under load), there might be an argument that responding to failed
messages increases the pressure.

## Security considerations

Sending a custom regex with an event might open the doors for attacking a
homeserver and/or a client by exposing a direct pathway to the complex code of a
regex parser. Additionally sending arbitrary complex regexes might make Matrix
more vulnerable to DoS attacks. To mitigate these risks it might be sensible to
only allow a more restricted subset of regular expressions by e.g. requiring a

> maximal length or falling back to simple globbing.

Reviewer 4: This should be worked out in this MSC, as we'll state in the spec whether a field supports full regex or only simple globbing.

@Half-Shot would a bridge ever need more than globbing for calling out affected users? Currently application service registration allows for full regex parsing (https://matrix.org/docs/spec/application_service/unstable#registration). But this is on the bridge side, and thus if it kills the homeserver, it was the homeserver operator that was at fault for using a bad registration file. Things are entirely different from the C-S API side.

Author: If using globbing, there would be a conversion needed from the AS regex, which it should be based on. As the regex language is more powerful than globbing, some simplifications/hacks/heuristics are required there. Or having the bridge user add it in two different forms manually…

sending the message that was presented him as failed; the bridge should have the
ability to mark an error as being revoked.

If [MSC 1410: Rich
Bridging](https://github.com/matrix-org/matrix-doc/issues/1410) is utilized for
this proposal it would additionally give the benefits of

- trimming the number of properties required in each bridge error event by
  separately providing these general infos about the bridge in the room state instead.
- not requiring users representing the bridge to have admin power levels
  (see [Rights management](#rights-management)).

### Bridge error event

This document proposes the addition of a new room event with type
`m.bridge_error`. It is sent by the bridge and references an event previously
sent in the same room, by that marking the original event as “failed to deliver”
for all users of a bridge. The new event type utilizes reference aggregations
([MSC
1849](https://github.com/matrix-org/matrix-doc/blob/matthew/msc1849/proposals/1849-aggregations.md#relation-types))
to establish the relation to the event its delivery it is marking as failed.
There is no need for a new endpoint as the existing `/send` endpoint will be
utilized.

> Additional information contained in the event are the name of the bridged

Reviewer 2: We can actually drop this now, when #2346 gets merged :)

Bridging](https://github.com/matrix-org/matrix-doc/issues/1410) is adopted. In
this case the same information is stored alongside other bridge metadata in the
room state*

There are some common reasons why an error occurred. These are encoded in the
`reason` attribute and can contain the following types:

* `m.event_not_handled` Generic error type for when an event can not be handled
  by the bridge. It is used as a fallback when there is no other more specific
  reason.

* `m.event_too_old` A message will – with enough time passed – fall out of its
  original context. In this case the bridge might decide that the event is too
  old and emit this error.

* `m.foreign_network_error` The bridge was doing its job fine, but the foreign
  network permanently refused to handle the event.

* `m.unknown_event` The bridge is not able to handle events of this type. It is
  totally legitimate to “handle” an event by doing nothing and not throwing this
  error. It is at the discretion of the bridge author to find a good balance
  between informing the user and preventing unnecessary spam. Throwing this
  error only for some subtypes of an event is fine.

> * `m.bridge_unavailable` The homeserver couldn't reach the bridge.

Reviewer: as a subclass to this which has shown to be problematic in recent weeks: When the homeserver is also dead, the users on other homeservers will see the message as delivered when in fact it is not.

I don't know if it makes total sense here given the traffic concern, but maybe flipping this proposal around for positive reactions to messages when they are delivered? Maybe a new kind of `m.receipt` for this specific purpose.

or maybe we train the general public that the bridge sending a read receipt is fine?

Presumably these ideas have already been covered, so I'm curious as to what the decisions were that led to it not being used.

* `m.no_permission` The bridge wanted to handle an event, but didn't have the
  permission to do so.

> The bridge error can provide a `time_to_permanent` field. If this field is

Reviewer: or `permanent: true` instead?

other foreign networks it bridges to. Matrix with its eventual consistency is
unique in having a message delivery guarantee. Because of this property there is
no need in the Matrix network itself to model the failure of message delivery.
This need only arises for interactions with foreign networks where message
delivery might fail. This proposal extends Matrix to be aware of these error
cases.

Additionally there might be some operational restrictions of bridges which might
make it necessary for them to refrain from handling an event, e.g. when hitting
memory limits. In this case the new event type can be used as well.

This is an example of how the new bridge error might look:

```
{
    "type": "m.bridge_error",
    "content": {
        "network: "Discord",
        "affected_users": ["@discord_.*:example.org"],
        "reason": "m.bridge_unavailable",
        "time_to_permanent": 900,
        "m.relationship": {
            "rel_type": "m.reference",
            "event_id": "$some:event.id"
        }
    }
}
```

\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\
¹ Or similar – see [Security Considerations](#security-considerations)

> ### Retries and error revocation

Reviewer: Retries might be better suited for a dedicated MSC given the complexity.

moment a transaction delivery to the bridge failed. The clients at this point
will start showing an error. When the bridge comes back online it will encounter
a higher-than-normal load as all events accumulated over the downtime are
flooding in. To handle this scenario well, the bridge will want to simply
discard all messages older than a given threshold and not bother with sending
any answer back.

By including a timeout in the `time_to_permanent` field of the event, the client
will know without further feedback from the homeserver or bridge when the
message won't be delivered anymore.

For those events still accepted by the bridge, the error must be revoked by a
`m.bridge_error_revoke` as described in the previous chapter.

**Note:** For this to work, the homeserver is required to impersonate a user of
the bridge as it has no agent of its own. The impersonated user would be the
bridge bot user or one of the virtual users in the bridge's namespace.

### Rights management

Only bridges should be allowed to send bridge errors and revocations.

Utilizing the rights system of the room provides a good approximation to this
behavior. It is fine to use it under the assumptions that

> - `m.bridge_error` and `m.bridge_error_revoke` require admin power levels.

Reviewer: by default they don't, unless you are expecting this to go into a whole new room version (which is a much harder sell)

> - there is always the bridge bot user or a virtual user in the bridge's

Reviewer: clients cannot check this

  namespace present in the room.
- at least one of those users possesses admin power level.
- all users with admin power levels are trusted.

In short, this requires giving bridges admin power levels in a room and trusting
them to restrict their actions to their own business. It is enough to have one
privileged bridge user in the room. In public rooms this is most commonly the

> bridge bot user with admin power level available and in 1:1 conversations it is

Reviewer: More commonly the bridge does *not* have any kind of power in the room. When bridges are admins, they are often added through Scalar which makes this decision for them - the bridges themselves do not acquire power to operate. There's also plenty of bridges which are not represented in Scalar, which has lead to a majority of rooms not having appropriate permissions for all bridges.

* `m.unknown_event` The bridge is not able to handle events of this type.

* `m.bridge_unavailable` The homeserver couldn't reach the bridge.

> * `m.no_permission` The bridge wanted to react to an event, but didn't have

Reviewer 2: I feel like we should use a word other than react, just because it's a specific action in itself (:+1)

Reviewer 2: The word I was after is "handle"
